### PR TITLE
Implement an interface for HDF5 hyperslabs.

### PR DESCRIPTION
**Pull request by 1uc**
This implements an API for HDF5 hyperslabs. It also reimplements existing selection functionality using the new wrappers.

### CODE REVIEW

**Review by alkino — APPROVED** [review]

File: include/highfive/bits/H5Slice_traits.hpp [hunk 1]
```diff
@@ -49,10 +49,239 @@ class ElementSet {
...
+        for (const auto& sel : selects) {
+            if (sel.op == Op::None) {
+                H5Sselect_none(space.getId());
+            } else {
+                if (H5Sselect_hyperslab(space.getId(), convert(sel.op),
+                                        sel.offset.empty() ? nullptr : sel.offset.data(),
+                                        sel.stride.empty() ? nullptr : sel.stride.data(),
+                                        sel.count.empty() ? nullptr : sel.count.data(),
+                                        sel.block.empty() ? nullptr : sel.block.data()) <
+                    0) {
```
Comment (matz-e) [inline include/highfive/bits/H5Slice_traits.hpp:216]: This looks funky

File: include/highfive/bits/H5Slice_traits.hpp [hunk 2]
```diff
@@ -49,10 +49,239 @@ class ElementSet {
...
+                    HDF5ErrMapper::ToException<DataSpaceException>(
+                        "Unable to select hyperslab");
+                }
+            }
+        }
+        return space;
+    }
+
+  private:
+    enum Op {
```
Comment (matz-e) [inline include/highfive/bits/H5Slice_traits.hpp:226]: Should this be an `enum class`?

File: include/highfive/bits/H5Slice_traits.hpp [hunk 3]
```diff
@@ -49,10 +49,239 @@ class ElementSet {
...
+        case NotA:
+            return H5S_SELECT_NOTA;
+        case Append:
+            return H5S_SELECT_APPEND;
+        case Prepend:
+            return H5S_SELECT_PREPEND;
+        case Invalid:
+            return H5S_SELECT_INVALID;
+        default:
+            return H5S_SELECT_INVALID;
```
Comment (matz-e) [inline include/highfive/bits/H5Slice_traits.hpp:263]: Seems like here we could throw an exception.

File: include/highfive/bits/H5Slice_traits.hpp (hunk 1 above)
Comment (1uc) [reply include/highfive/bits/H5Slice_traits.hpp:216]: Fixed in #542

File: include/highfive/bits/H5Slice_traits.hpp (hunk 3 above)
Comment (1uc) [reply include/highfive/bits/H5Slice_traits.hpp:263]: Fixed in #542

File: include/highfive/bits/H5Slice_traits.hpp (hunk 2 above)
Comment (1uc) [reply include/highfive/bits/H5Slice_traits.hpp:226]: Fixed in #542

**Pull request merged**